File: Sources/XMTP/Messages/UnsignedPublicKey.swift
//
//  UnsignedPublicKey.swift
//
//
//  Created by Pat Nakajima on 11/17/22.
//

import Foundation
import XMTPProto

typealias UnsignedPublicKey = Xmtp_MessageContents_UnsignedPublicKey

extension UnsignedPublicKey {
	static func generate() throws -> UnsignedPublicKey {
		var unsigned = UnsignedPublicKey()
		let key = try PrivateKey.generate()
		let createdNs = Date().millisecondsSinceEpoch
		unsigned.secp256K1Uncompressed.bytes = key.publicKey.secp256K1Uncompressed.bytes
		unsigned.createdNs = UInt64(createdNs)
		return unsigned
	}

	init(_ publicKey: PublicKey) {
		self.init()

		createdNs = publicKey.timestamp
		secp256K1Uncompressed.bytes = publicKey.secp256K1Uncompressed.bytes
	}
}

File: library/src/main/java/org/xmtp/android/library/messages/UnsignedPublicKey.kt
package org.xmtp.android.library.messages

import java.util.Date

typealias UnsignedPublicKey = org.xmtp.proto.message.contents.PublicKeyOuterClass.UnsignedPublicKey

fun UnsignedPublicKey.generate(): UnsignedPublicKey {
    val unsigned = UnsignedPublicKey.newBuilder()
    val key = PrivateKey.newBuilder().build().generate()
    val createdNs = (Date().time * 1_000_000)
    unsigned.secp256K1Uncompressed = unsigned.secp256K1Uncompressed.toBuilder().also {
        it.bytes = key.publicKey.secp256K1Uncompressed.bytes
    }.build()
    unsigned.createdNs = createdNs
    return unsigned.build()
}

class UnsignedPublicKeyBuilder {
    companion object {
        fun buildFromPublicKey(publicKey: PublicKey): UnsignedPublicKey {
            return UnsignedPublicKey.newBuilder().apply {
                createdNs = publicKey.timestamp
                secp256K1Uncompressed = secp256K1Uncompressed.toBuilder().also {
                    it.bytes = publicKey.secp256K1Uncompressed.bytes
                }.build()
            }.build()
        }
    }
}
